src/pybind/mgr/influx/module.py

```diff
@@ -265,70 +353,78 @@ def send_to_influx(self):
                               "manually and grant write privs to user "
                               "'%s'", self.config['database'],
-                              self.config['username'])
+                              self.config['database'])
                 client.create_database(self.config['database'])
-                client.create_retention_policy(name='8_weeks', duration='8w',
-                                               replication='1', default=True,
+                client.create_retention_policy(name='8_weeks',
+                                               duration='8w',
+                                               replication='1',
+                                               default=True,
                                                database=self.config['database'])
-            else:
-                self.set_health_checks({
-                    'MGR_INFLUX_SEND_FAILED': {
-                        'severity': 'warning',
-                        'summary': 'Failed to send data to InfluxDB',
-                        'detail': [str(e)]
-                    }
-                })
-                raise
+            client.close()
+
+            self.log.debug('Gathering statistics')
+            points = self.gather_statistics()
+            for chunk in self.chunk(points, self.config['batch_size']):
+                self.queue.put(chunk, block=False)
+
+            self.log.debug('Queue currently contains %d items',
+                           self.queue.qsize())
+        except queue.Full:
+            self.health_checks.update({
+                'MGR_INFLUX_QUEUE_FULL': {
+                    'severity': 'warning',
+                    'summary': 'Failed to chunk to InfluxDB Queue',
+                    'detail': ['Queue is full. InfluxDB might be slow with '
+                               'processing data']
+                }
+            })
+            self.log.error('Queue is full, failed to add chunk')
+        except (RequestException, InfluxDBClientError) as e:
+            self.health_checks.update({
+                'MGR_INFLUX_DB_LIST_FAILED': {
+                    'severity': 'warning',
+                    'summary': 'Failed to list/create InfluxDB database',
+                    'detail': [str(e)]
+                }
+            })
+            self.log.exception('Failed to list/create InfluxDB database')
+            return False
+        finally:
+            self.set_health_checks(self.health_checks)
 
     def shutdown(self):
         self.log.info('Stopping influx module')
         self.run = False
         self.event.set()
+        self.log.debug('Shutting down queue workers')
+
+        for _ in self.workers:
+            self.queue.put(None)
+
+        self.queue.join()
+
+        for worker in self.workers:
+            worker.join()
 
     def self_test(self):
-        daemon_stats = self.get_daemon_stats()
+        now = self.get_timestamp()
+        daemon_stats = self.get_daemon_stats(now)
```

Comment:
`self.get_daemon_stats(now)` now returns an iterator which can not be `len()`'ed


```
2018-08-21T10:06:01.202 INFO:tasks.ceph.mgr.x.mira001.stderr:2018-08-21 10:06:01.202 7f12b01ad700 -1 Remote method threw exception: Traceback (most recent call last):
2018-08-21T10:06:01.203 INFO:tasks.ceph.mgr.x.mira001.stderr:  File "/usr/lib/ceph/mgr/influx/module.py", line 426, in self_test
2018-08-21T10:06:01.203 INFO:tasks.ceph.mgr.x.mira001.stderr:    assert len(daemon_stats)
2018-08-21T10:06:01.203 INFO:tasks.ceph.mgr.x.mira001.stderr:TypeError: object of type 'generator' has no len()
2018-08-21T10:06:01.203 INFO:tasks.ceph.mgr.x.mira001.stderr:
```